src/util/walkStrategy.ts

```diff
@@ -0,0 +1,123 @@
+import { BaseTrie } from '..'
+import { FoundNodeFunction } from '../baseTrie'
+import { PrioritizedTaskExecutor } from '../prioritizedTaskExecutor'
+import { BranchNode, ExtensionNode, LeafNode, Nibbles, TrieNode } from '../trieNode'
+
+export default class WalkStrategy {
+  readonly onNode: FoundNodeFunction
+  readonly taskExecutor: PrioritizedTaskExecutor
+  readonly trie: BaseTrie
+  private resolve: Function
+
+  /**
+   * Creates a new WalkStrategy
+   * @param onNode - The `FoundNodeFunction` to call if a node is found
+   * @param trie - The `Trie` to walk on
+   */
+  private constructor(onNode: FoundNodeFunction, trie: BaseTrie, poolSize: number) {
+    this.onNode = onNode
+    this.taskExecutor = new PrioritizedTaskExecutor(poolSize)
+    this.trie = trie
+    this.resolve = () => {}
+  }
+
+  /**
+   * Async function to create and start a new walk over a trie.
+   * @param onNode - The `FoundNodeFunction to call if a node is found
+   * @param trie - The trie to walk on
+   * @param root - The root key to walk on
+   * @param poolSize - Task execution pool size to prevent OOM errors. Defaults to 500.
+   */
+  static async newWalk(
+    onNode: FoundNodeFunction,
+    trie: BaseTrie,
+    root: Buffer,
+    poolSize?: number
+  ): Promise<void> {
+    const strategy = new WalkStrategy(onNode, trie, poolSize ?? 500)
+    await strategy.startWalk(root)
+  }
+
+  startWalk(root: Buffer): Promise<void> {
+    return new Promise(async (resolve) => {
+      this.resolve = resolve
+      const node = await this.trie._lookupNode(root)
+      this.processNode(root, node as TrieNode, [])
+    })
+  }
+
+  /**
+   * Run all children of a node. Priority of these nodes are the key length of the children
+   * @param node - Node to get all children of and call onNode on
+   * @param key - The current `key` which would yield the `node` when trying to get this node with a `get` operation.
+   */
+  allChildren(node: TrieNode, key: Nibbles = []) {
+    if (node instanceof LeafNode) {
+      if (this.taskExecutor.finished()) {
```

Comment:
This should probably be a global condition, need to figure out where we want to check this